### ISSUE TITLE
Decision to be taken on Raku filename extensions.

### ISSUE THREAD

**Issue by vrurg** [perl6/problem-solving#106]
It is to be decided what filename extensions are to be used for Raku if language renaming happens.

**Comment by vrurg** [comment]
This is a followup to #101 where we only decide what extensions are to be used. Proposals on using same extension for all file types are not welcome.

The bottom line of #101 is following:

The base (script) extension is to be chosen of two options: wether it be *.raku* or *.rk*. Depending on this accompanying extensions could be:

- for *.raku*
   - *.rakulib* or *.rakumod* – TBD
   - *.rakutest* – TBD, I didn't find any good option for tests in #101 
   - *.rakudoc* - perhaps, everybody would agree
- for *.rk*
   - *.rkm* 
   - *.rkd*
   - no good option for test files.


- *.raku* pros and cons:
   - *pros*
      - guaranteed uniqueness and better recognizability.
      - less chances of concurring with other file types for editor support
   - *cons*
     - lengthy
- *.rk*
   - *pros*
      - shortness
   - *cons*
      - no good option for test files. The best one would be *.rkt* but it is taken by Racket.

**Comment by vrurg** [comment]
My opinion on test extension is that the only people who could propose a radical change are those who develops testing tools and modules. @Leont, in particular.

**Comment by ugexe** [comment]
I still don't understand the need for this. We did not create a separate issue to vote on the name or other minor details -- one comes with the other and is thus part of it.

**Comment by AlexDaniel** [comment]
@ugexe I'm not following your point. This is just a ticket to have a dedicated discussion about extensions without cluttering the existing PR (which is already hard to follow).

**Comment by vrurg** [comment]
@ugexe based on [Jonathan's comment](https://github.com/perl6/problem-solving/issues/101#issuecomment-530144263), the decision would be made once and forever. The only other one-way decision is the rename itself. Looking at #101 we may conclude that the extension subject is rather controversial. The best way to resolve it is to vote.

**Comment by ugexe** [comment]
It appears controversial because it is incredibly bikesheddable. And since this is literally *part of the other PR* it belongs there.

**Comment by AlexDaniel** [comment]
This is a thread to discuss the extensions, so that we can adjust the main PR. Please don't bring unnecessary pessimism here.

**Comment by ugexe** [comment]
Yes, leave pessimism at the door. Wouldn't want anyone going against the grain 🙄

**Comment by AlexDaniel** [comment]
@ugexe in this discussion you're not going against the grain.

**Comment by ugexe** [comment]
@AlexDaniel in this discussion you're not convincing me of anything.

**Comment by vrurg** [comment]
@ugexe the whole point of this topic is just to vote. Get a minor decision on wether it be *rakumod* or *rakulib*, two other similar decisions – and choose. It would turn into bikeshedding if we chose to do so. Or we not.

BTW, it's rather for you to decide impartially on which of the two long module extensions is better.

**Comment by ugexe** [comment]
Bike shedding is the tendency of unimportant issues to receive undue attention because the barrier of entry to give an opinion on the subject is extremely low or non-existent.

**Comment by AlexDaniel** [comment]
OK, this ticket is for discussing the filename extensions. We need to choose some good extensions because https://github.com/perl6/problem-solving/pull/89 is currently blocked by it. It is *good* to have a separate ticket because following discussions on a PR was proven to be difficult. Once we have enough info here, the main PR will be adjusted.

Please file a separate ticket for any other matter.

**Comment by AlexDaniel** [comment]
Hypothetically, it's also possible to have both short and long versions. It's not uncommon, for example `.markdown` and `.md`. This does create a problem if a module has both `Foo.rkm` and `Foo.rakumod` files (so which one has the priority?), but at the same time it's unlikely that a module author would do something like this, so maybe it's not that big of an issue (after all we did have both `.pm` and `.pm6`, and while it's painful, it does sorta work).

**Comment by tony-o** [comment]
Move forward with the `.raku[mod|doc|test]` extensions and let's move on.

**Comment by hythm7** [comment]
My opinion on this is:
`.raku` for modules.
No extension for scripts and tests. ( unless it is required for MS Windows, Then use same extension  `.raku` ).

**Comment by AlexDaniel** [comment]
@hythm7 you're free to use no extension for scripts, but it seems to be needed. Also, as previously discussed, we're not going to use the same extension for scripts and modules.

**Comment by vrurg** [comment]
As I don't want this to take too long and turn into another round of bikeshedding, let do it this way:

- let's not discuss but vote
- my next three comments would represent three options to choose of
- the options will contain final and non-discussable variants of extensions. The only changes allowed are:
    - for modules: by @ugexe
    - for test files: by @Leont
    - by @jnthn to any except for the base extensions (or otherwise this all wouldn't make any sense)
- like (the thumb up) means a vote for an option. No other "like" is taken into account.
- please, don't vote two options at once. Such votes could be excluded from total count.
- the voting is in effect for three days with the first day starting on Sep, 17. It Europe it is effectively "today" now, as I'm writing this
- the voting result is to be considered by @jnthn and it is up to him to decide wether to change the renaming PR with accordance to the public opinion or not

**Comment by vrurg** [comment]
### Option 1. *.raku*

*.rakumod*, *.rakudoc*, *.rakutest*

**Comment by vrurg** [comment]
### Option 2. *.rk*

*.rkm*, *.rkd*, *.t*

Test file could be reconsidered later, as it already happened with *.t6*.

**Comment by vrurg** [comment]
### Option 3. 1+2

Both option 1 and 2 to be supported as proposed by @AlexDaniel above.

**Comment by japhb** [comment]
Sorry, I'm not trying to bikeshed, but for option #2, where was the decision not to use `.rt`?  I thought I saw someone mention that one, but then I lost track of where it was decided against.  (I personally felt it was a reasonable compromise to avoid stomping on `.rkt`, but again I'm not trying to bikeshed, just to figure out where I got lost here.)

**Comment by vrurg** [comment]
@japhb the discussion was a bit scattered over a couple of places. *.rt* stands for RealTime Player files. Could though be considered with care because it would barely be a file format known to editors and thus won't cause conflicts. But I'm not sure about it, anyway.

**Comment by AlexDaniel** [comment]
@japhb it's a fair question. @jnthn wanted something that has a prefix (like “.rk*”), which is why we started to drift away from “.rd” for docs. But `.rkt` is a real problem so I'm not sure if this idea is really useful…

**Comment by duncand** [comment]
In my opinion, the extra few letters in the longer extensions are a complete non-problem, when you consider how often they are typed and how long the rest of the filename/path/etc they are part of already are, so the `raku` form wins in every important metric.

**Comment by vrurg** [comment]
BTW, we played a game of "imagine your file with new extensions" on IRC. Perhaps it worth visualizing how things would look:

```
bin/myscript.raku
lib/MyModule.rakumod
doc/MyModule.rakudoc
t/00-test.rakutest
```

```
bin/myscript.rk
lib/MyModule.rkm
doc/MyModule.rkd
t/00-test.t
t/01-test.rt # Ok, what if...
```

**Comment by duncand** [comment]
As much as I strongly prefer `raku` I believe there is a way out of the `rk` problem with tests that arises if you don't strictly require a single letter.  For example, while somewhat imbalanced, a way out is `rktest` for tests while the suggested `rk/rkm/rkd` for the other things.  Just putting that out there.  I still want the `raku` base though.

**Comment by shishini** [comment]
for the .rk option
may i suggest .rktst for the test file 
its .rakutest - all the voyelles (a,e,i,o,u,y)

**Comment by duncand** [comment]
I believe the `rk` option can avoid conflicts with Racket et al if it isn't strictly limited to `rk` plus 1 letter, eg if 2 letters is ok.  To be specific, I suggest `rktt` for tests which can also evoke the `t/*.t` of old.  It shouldn't be a problem to have say `rk/rkm/rkd/rktt` if that's all it is.  I still prefer the `raku` stream but a change like this can make the `rk` stream easier to live with.

**Comment by vrurg** [comment]
For the record: I'm not going to change the options except for the situations stated in the voting rules. But any reasonable proposal expressed in comments could be considered later.

`.rktt` looks promising to me. Though I didn't check yet if it's used somewhere.

**Comment by duncand** [comment]
My basic Google search found nothing using `.rktt`.

As for changing the existing options, I expected you wouldn't but I'm assuming the 3 named individuals in the voting rules might do so.

**Comment by lizmat** [comment]
To prevent *any* ambiguity, we need to move our thinking out of the 8:3 mindset:  I therefore strongly support @tony-o:

> Move forward with the .raku[mod|doc|test] extensions and let's move on.

**Comment by duncand** [comment]
Adding your thumbs up vote to Option 1 will make more sure that support gets counted.

**Comment by duncand** [comment]
What @jnthn said has some appeal, and may make some wonder why `ra` wasn't even mentioned before in favor of `rk`.

However, 1 big issue I see with those options is that `ra` and `ram` etc were widely used multimedia file formats connected with RealPlayer, and it is very likely that various modern multimedia programs like VLC will try and play them.

This is considerably worse than `rk` being an old archive file format; the RealPlayer files are likely to be quite widely distributed still.

I still think `raku` is much better extension base.

**Comment by lizmat** [comment]
FWIW, on MacOS `.ram` is indeed associated with RealPlayer, and attempts to play it with RealPlayer when opening it.  Although MacOS doesn't come with a RealPlayer.

**Comment by lizmat** [comment]
> I still think raku is much better extension base.

Which would make `.rakut`, which is at least in Dutch, weird, and sounds like Racket.

**Comment by duncand** [comment]
Just throwing this out there: `rakup`, `rakum`, `rakud`, `rakut`.

**Comment by duncand** [comment]
The "interface" seems too specific for what the file contains, the `ral` for library is more comprehensive on what the files would contain.

**Comment by vrurg** [comment]
@jnthn Shall I add *Option 4* and restart the vote? Though RealTime associations put me very much uncertain about *.ra* family.

**Comment by AlexDaniel** [comment]
People should feel free to change their reactions. This is not exactly the same voting that we do in the problem-solving repo anyway, I see reactions as merely an indication of which extensions people will prefer to use. I think @jnhtn and @lizmat will decide how to adjust the main PR.

**Comment by vrurg** [comment]
@AlexDaniel the point is to make clear it is a 4th option. Wouldn't it be bound to RealTime I'd add it already.

**Comment by borisdaeppen** [comment]
The `.rk`-line always felt so "klingon" to me (meaning "hard", "rough", "tongue-twister").
Both for the eyes (a lot of sharp corners) and mouth (only hard  consonants).

Just because of this I'm positively surprised about the `.ra` idea.

**Comment by tmtvl** [comment]
.rat is already used for PICS Web rating system implementations, though they are not very popular for W3C standards.
.rad is used by the Source engine and files are formatted as XML.

**Comment by duncand** [comment]
@jnthn @vrurg There are other subsets of Raku we didn't consider yet.

May I propose, `ru` (for `RakU`) as the base, assuming there isn't some negative connotation.

From what I saw, only `run` is already used by something; `ru` itself is also available.

So then we have `rup`, `rum`/`rul`, `rud`, `rut` for extensions.

Alternatively, `rl*` for `Raku Language`.

So then we have `rlp`, `rlm`/`rll`, `rld`, `rlt` for extensions.

I suggest that https://www.reviversoft.com/file-extensions/?char=r is useful at a glance about options, assuming it is reasonably thorough.

Any thoughts on that?

**Comment by duncand** [comment]
Another variation is `rak` (`RAKu`) as the common base.  Then we can use `raku` for main programs and also `rakm`/`rakl`, `rakd`, `rakt` for others.  Personally this looks kind of like pirate talk, eg "red rackhams treasure".

**Comment by vrurg** [comment]
@duncand *.ru* and *.rak* are both rather far-fetched. Human brain won't really bind *ru* to *Raku* and thus won't be easily creating associative link. *rak* is just one letter away from *raku* which also makes little sense in choosing it.

**Comment by vrurg** [comment]
For test files I'm permanently coming back to `.t.rk` variant – a slight development of `_test` suffix proposed in #101:

```
00-mytest.t.rk
```

It doesn't really makes me all happy, but it serves for two purposes at once: it clearly identifies file type; and it clearly declares the file to be a test.

**Comment by niner** [comment]
I very much like the .t.rk or .t.whatever-extension-we-chose-for-programs idea. It resolves the ambiguity with .t files and eliminates the chance of an oversight in tools which may miss the test file extension while still marking test files as such.

**Comment by Mekk** [comment]
1. What about .trk (memo: test rk) for tests?

2. Of funny note, .rk is somehow related to .pl (r follows p, k precedes l – at least in alphabets where q is not used)

**Comment by vrurg** [comment]
@Mekk *.trk* is in use by GPS tracking data.

**Comment by scriplit** [comment]
Perl 6 invented twigils, nobody would be surprised if Raku would now introduce twilexts ;)
bin/myscript.raku
lib/MyModule.m.raku
doc/MyModule.d.raku
t/00-test.t.raku

**Comment by vrurg** [comment]
@scriplit that was the second thing on my mind, but I rejected it because it is unlikely to be welcomed by some. Anyway, if we go this way, I would rather consider `.rk` then and have:

```
bin/myscript.rk
lib/MyModule.m.rk
doc/MyModule.d.rk
t/00-test.t.rk
```

Also *twextensions* would ease adding new file types in the future if needed. Though they would have to be syntax-compatible with Raku itself as to not drive crazy editors which rely on the last extension exclusively – if there're any of the kind.

**Comment by duncand** [comment]
Oh, please, can we do that?  Twexts for everyone!

The main extension indicates the actual format of the file and how to parse/interpret it (the Raku grammar), which is the same for all the use cases we've been discussing (and why I originally proposed a single unified extension).

The sub-extension provides more semantics on the main purpose of the file (app/lib/doc/sub).

This approach means we minimize actual footprint and combinations.

Exploiting this, I recommend we reserve BOTH `rk` and `raku` main extensions and let people use what they want similar to the md/markdown dual; each one can share the same sub-extensions.

I also recommend for parity that we have an official sub-extension for main programs, eg `p`/`a` for program/app, but the sub-extension is optional and implicitly means main if absent.

**Comment by tony-o** [comment]
> (and why I originally proposed a single unified extension).

As discussed ad nauseum, a single unified extension is a headache that's not well thought through.  @vrurg can we please lock this and start a vote?  Closing the voting and reopening it is not going to get a decent vote (a lot of people have already voted).  Further, if this is really holding up the referenced PR this bike shedding is not useful and holding up more important tasks.

**Comment by duncand** [comment]
@tony-o I'm not proposing a return to a unified extension. Practically speaking the twext option is still multiple extensions.

**Comment by duncand** [comment]
@tony-o The voting already started yesterday and never stopped.  And your proposal is winning with 10 thumbs-up votes.  This discussion is on the side and doesn't take away from the voting.

**Comment by vrurg** [comment]
> @vrurg can we please lock this and start a vote? 

The voting is currently active. Because it is has rather advisory status (meaning it's results are not decisive and only provide info for @lizmat and @jnthn to make up their mind), there is no point in blocking discussion because it might result in some fresh new ideas to be taken into account.

In either case, Jun 19 is the last voting day when it all ends and we proceed with the main PR.

**Comment by nxadm** [comment]
> ### Option 1
> 
> _.raku_, _.rakumod_, _.rakudoc_, _.rakutest_

What about having extensions of the same length for non script files? Shells and other programs using fixed width fonts will look a lot prettier...
```
foo.raku
foo.rakulib
foo.radkudoc
foo.rakutst
```

**Comment by AlexDaniel** [comment]
> Shells and other programs using fixed width fonts will look a lot prettier...

But why does that matter? Strings of different length are easier to distinguish.

**Comment by nxadm** [comment]
> > Shells and other programs using fixed width fonts will look a lot prettier...
> 
> But why does that matter? Strings of different length are easier to distinguish.

I think you underestimate the people with a soft form of OCD in programmer's circles.

**Comment by AlexDaniel** [comment]
> soft form of OCD

Should we care? And what about mild forms of impaired vision :P

Anyway, test files are unlikely to be near other files often, and if that happens I don't mind the other files to stand out a bit.

**Comment by nxadm** [comment]
> > soft form of OCD
> 
> Should we care? And what about mild forms of impaired vision :P
> 
> Anyway, test files are unlikely to be near other files often, and if that happens I don't mind the other files to stand out a bit.

And these?

 ```
foo.raku
foo.t.raku
foo.rakulib
foo.radudoc
```
Too incoherent? Just right?

**Comment by vrurg** [comment]
@nxadm I think "incoherent" is the right word for the case.

**Comment by vrurg** [comment]
For the record: the voting ends today.

**Comment by nxadm** [comment]
(@vrurg, don't forget about time zones...)

**Comment by vrurg** [comment]
@nxadm I remember. I think it would be ok to close at 00:00 PST. Though considering the current activity, it's unlikely to see any more votes.

**Comment by nxadm** [comment]
Yes, just to be sure. It looks like who wanted to vote on this issue already did. (I abstained because I am fine with whatever is chosen.)

**Comment by vrurg** [comment]
The voting is over. The bottom line:

Option 1: 11
Option 2: 2
Option 3: 7
Option 4: 5

Interestingly enough, twextensions got 6 votes even though they were not on the list. Would they be there my vote would go for them too.

Anyway, I think we could say that community has made its choice and we can proceed further with #89.

**Comment by duncand** [comment]
As a status update, Liz has just updated the pull request https://github.com/perl6/problem-solving/pull/89 to use the raku/rakumod/rakudoc/rakutext options, thus resolving this issue.
